runtime/signals.c

```diff
@@ -33,21 +33,22 @@
 #include "caml/memprof.h"
 #include "caml/finalise.h"
 
-#ifndef NSIG
-#define NSIG 64
-#endif
+/* The set of pending signals (received but not yet processed).
+   It is represented as a bit vector.
+   Valid signal numbers range from 1 to NSIG - 1 included.
+   (This is checked when we install a signal handler.)
+   Signal 1 is the least significant bit of caml_pending_signals[0]. */
 
-/* The set of pending signals (received but not yet processed) */
+CAMLexport atomic_uintnat caml_pending_signals[NSIG_WORDS];
 
-CAMLexport atomic_intnat caml_pending_signals[NSIG];
 static caml_plat_mutex signal_install_mutex = CAML_PLAT_MUTEX_INITIALIZER;
 
 int caml_check_for_pending_signals(void)
 {
   int i;
 
-  for (i = 0; i < NSIG; i++) {
-    if (atomic_load_explicit(&caml_pending_signals[i], memory_order_seq_cst))
+  for (i = 0; i < NSIG_WORDS; i++) {
+    if (atomic_load_explicit(&caml_pending_signals[i], memory_order_relaxed))
```

Comment:
One particular ordering to think about: when a signal is recorded, `caml_pending_signals` is set atomically and then `young_limit` is set with a release store. This message is received when the comparison `young_limit <= young_pointer` fails, after which `caml_check_for_pending_signals` ends up being called via `caml_process_pending_signals`.

Now the comparison with `young_limit` is done without synchronisation, so `caml_pending_signals` could still be seen as unset after the comparison fails. I am not sure this actually happens currently, but only for accidental reasons (an unrelated acquire load of `young_limit` in `handle_gc_interrupts` which happens to be called before `caml_process_pending_signals` inside `caml_process_pending_actions`, and whole lot of other work happen between the two; even so I am not sure how to reason about it).

It would be clearer and more robust to change the comparison pattern to upgrade the relaxed load into an acquire load when it fails using a fence, e.g.
```
if (young_pointer < young_limit) {
  atomic_thread_fence(memory_order_acquire);
  …
}
```
(the code in old trunk was organised in such a way that there is a common entry point to the GC which would be good to bring back, in this case it is an obvious place for such a fence).

Reply:
As far as this PR is concerned one solution is to add this fence at the top `caml_check_for_pending_signals`, but I can move this comment to my review of the multicore PR if you want. (I already commented there about the choice of a volatile read of young_limit instead of an atomic relaxed read which is UB.)

Reply:
We could put an "acquire" fence at the beginning of `caml_check_for_pending_signals`, or use "acquire" loads instead of "relaxed" loads.  But it's not clear to me it's doing much of a difference.  There will always be races that cause a pending signal to not be processed immediately, no?

Since I'm clearly out of my league here, I'd appreciate experts to chime in and tell me what to do.  This PR speeds up the bytecode interpreter considerably, so I'd really like to see it merged soon.

Reply:
We're not doing an acquire on `young_limit`, which I think opens up the theoretical possibility of seeing `young_limit` updated without `caml_pending_signals` being set and then resetting `young_limit` after the check. This would delay processing the pending signal until `young_limit` is updated again (another signal, stop-the-world of some kind or we enter a blocking section). I think it might be possible to engineer a scenario where this leads to never processing a SIGINT (loop with a safepoint poll and nothing else?).

I think this can happen anyway at the moment. A signal could be delivered at any time before https://github.com/xavierleroy/ocaml/blob/79fb0b0ddaca1051106fcd540b5e82960449218d/runtime/minor_gc.c#L635 and the resulting young_limit would be over-written. I'll add that to the list of signalling issues, we probably need a `caml_check_for_pending_signals` at the very end of a stw section.

I don't think the memory orderings need hold up the PR, it doesn't worsen the existing situation and we can look to remedy things in the follow-up with the rest of @gadmm 's identified issues on the multicore PR.

Reply:
* I'm not sure it can happen at the moment in practice (series of seq_cst loads, the first of which is for signal 0). I agree it's a bit independent from this PR because I'd want the fence anyway so that correctness follows from the structure of the program.
* I'm fine with a fence at the top of `check_for_pending_signals` and we refine it later (I am not sure acquire loads here are a proper fix).
* Good point about missing signals anyways; I don't think so, because the former trunk implementation went to length to avoid that, by processing signals first for instance. @sadiqj points out an issue with the multicore implementation, which in old trunk was dealt with by the update_young_limit business. There is still the case of races with syscalls but the reason there is fundamental rather than a programming bug, and there was an idea to fix this too.

Reply:
> I'm fine with a fence at the top of check_for_pending_signals and we refine it later

OK, done.  I added a `[MM]` (memory model) comment briefly summarizing the discussion and pointing here.